sticker2/src/input/xlm_roberta.rs

```diff
@@ -0,0 +1,95 @@
+use conllx::graph::{Node, Sentence};
+use sentencepiece::SentencePieceProcessor;
+
+use crate::input::{SentenceWithPieces, Tokenize};
+
+const FAIRSEQ_BOS_ID: i64 = 0;
+const FAIRSEQ_EOS_ID: i64 = 2;
+const FAIRSEQ_OFFSET: i64 = 1;
+
+/// Tokenizer for Roberta models.
+///
+/// Roberta uses the sentencepiece tokenizer. However, we cannot use
+/// it in the intended way: we would have to detokenize sentences and
+/// it is not guaranteed that each token has a unique piece, which is
+/// required in sequence labeling. So instead, we use the tokenizer as
+/// a subword tokenizer.
+pub struct XLMRobertaTokenizer {
+    spp: SentencePieceProcessor,
+}
+
+impl XLMRobertaTokenizer {
+    pub fn new(spp: SentencePieceProcessor) -> Self {
+        XLMRobertaTokenizer { spp }
+    }
+}
+
+impl From<SentencePieceProcessor> for XLMRobertaTokenizer {
+    fn from(spp: SentencePieceProcessor) -> Self {
+        XLMRobertaTokenizer::new(spp)
+    }
+}
+
+impl Tokenize for XLMRobertaTokenizer {
+    fn tokenize(&self, sentence: Sentence) -> SentenceWithPieces {
+        // An average of three pieces per token ought to enough for
+        // everyone ;).
+        let mut pieces = Vec::with_capacity((sentence.len() - 1) * 3);
+        let mut token_offsets = Vec::with_capacity(sentence.len());
+
+        pieces.push(FAIRSEQ_BOS_ID);
+
+        for token in sentence.iter().filter_map(Node::token) {
+            token_offsets.push(pieces.len());
```

Comment:
This is what you meant here?
https://github.com/stickeritis/sticker-transformers/pull/30#issuecomment-582076105

Reply:
No, there I meant that the position indices in RoBERTa models start at 2. So, for a sentence of 5 non-padding tokens, we look up the position embeddings `[2, 3, 4, 5, 6]` rather than `[0, 1, 2, 3, 4]`. Padding should use index `1` which must be zero-initialized. But I don't think it matters in our case, since we are always masking padding timesteps anyway and we do not have padding in the middle of a sentence.

Reply:
Ah, I see. Then this here depends, if you manually increment the token offsets / cut off the first sentence bracket this should be fine. In my implementation I use pieces +1 as token offset to account for <s> which usually doesn't have a label.

That was a bug that I initially had which made things look pretty bad.

Reply:
Why do you add pieces +1? Or aren't you using `RobertaEmbeddings` from HF transformers? Since their implementation adds the necessary offset to the position embeddings.

Reply:
<s>I was referring to `token_offsets` which I suppose you are using to look up the predicted label for a token. Since we prepend the `<s>` bracket, the actual token indices are off by one:

Word-piece: `I am a te ##xt`
Offsets: 0,1,2,3

roberta: `<s> I am a te ##xt</s>`
Offsets: 1,2,3,4</s>

Nevermind, I saw the line above, you add the BOS token above, so `pieces.len()` has the right offset.